File: sdk/python/examples/01_guard.py
"""
Example: @clampd.guard() — Wrap any function with security.

Run: python examples/01_guard.py

To use a real gateway, remove the mock and set:
    CLAMPD_GATEWAY_URL=http://localhost:8080
    CLAMPD_API_KEY=your-key
"""
import sys

sys.path.insert(0, ".")

from examples.mock_gateway import install_mock_gateway

install_mock_gateway()

import clampd
from clampd.client import ClampdBlockedError

clampd.init(agent_id="guard-demo-agent")


@clampd.guard("db.query")
def query_database(sql: str) -> str:
    return f"Results for: {sql}"


@clampd.guard("drop_table")
def dangerous_delete(table: str) -> str:
    return f"Dropped {table}"


def main():
    # Allowed call
    print("--- Safe query ---")
    result = query_database(sql="SELECT name FROM users LIMIT 10")
    print(f"Result: {result}")

    # Blocked call
    print("\n--- Dangerous operation ---")
    try:
        dangerous_delete(table="users")
    except ClampdBlockedError as e:
        print("BLOCKED!")
        print(f"  Reason: {e.response.denial_reason}")
        print(f"  Risk: {e.response.risk_score}")


if __name__ == "__main__":
    main()

File: sdk/typescript/examples/01-guard.ts
/**
 * Example: clampd.guard() - Wrap any async function with security.
 *
 * Run: npx tsx examples/01-guard.ts
 *
 * To use a real gateway, remove the mock and set:
 *   CLAMPD_GATEWAY_URL=http://localhost:8080
 *   CLAMPD_API_KEY=your-key
 */
import { installMockGateway } from "./mock-gateway.js";
import clampd, { ClampdBlockedError } from "../src/index.js";

installMockGateway();

clampd.init({ agentId: "guard-demo-agent" });

// Your original tool function
async function queryDatabase(params: { sql: string }): Promise<string> {
  return `Results for: ${params.sql}`;
}

// Wrap it - every call is now guarded
const safeQuery = clampd.guard(queryDatabase, {
  toolName: "db.query",
  checkResponse: true,
});

async function main() {
  // Allowed call
  console.log("--- Safe query ---");
  const result = await safeQuery({ sql: "SELECT name FROM users LIMIT 10" });
  console.log("Result:", result);

  // Blocked call (drop_table is in the blocked list)
  console.log("\n--- Dangerous query ---");
  const dangerousQuery = clampd.guard(
    async (p: { table: string }) => `dropped ${p.table}`,
    { toolName: "drop_table" },
  );

  try {
    await dangerousQuery({ table: "users" });
  } catch (err) {
    if (err instanceof ClampdBlockedError) {
      console.log("BLOCKED!");
      console.log("  Reason:", err.response.denial_reason);
      console.log("  Risk:", err.response.risk_score);
      console.log("  Rules:", err.matchedRules);
    }
  }
}

main().catch(console.error);
